compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorDistribute.cpp

```diff
@@ -0,0 +1,343 @@
+    analysis.setAnchor(transfer.getResult(), layout);
+  }
+
+  ArrayAttr mmaTypes;
+  SmallVector<int64_t, 3> workgroupSize;
+
+  bool populatedMfma = false;
+  RewritePatternSet patterns;
+};
+
+struct LLVMGPUVectorDistributePass
+    : public LLVMGPUVectorDistributeBase<LLVMGPUVectorDistributePass> {
+public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::VectorExt::IREEVectorExtDialect>();
+    registry.insert<amdgpu::AMDGPUDialect>();
+    registry.insert<gpu::GPUDialect>();
+    registry.insert<nvgpu::NVGPUDialect>();
+  }
+
+  void runOnOperation() override {
+    auto func = getOperation();
+
+    FailureOr<ArrayAttr> maybeSupportedTypes =
+        getSupportedMmaTypes(llvm::cast<func::FuncOp>(func));
+    // TODO: Support FMA fallback. Contractions always benefit from an anchoring
+    // layout because they do implicit shuffles, or broadcast when loading data.
+    if (failed(maybeSupportedTypes)) {
+      func->emitError() << "Failed to collect the set of supported mma types "
+                           "for vector distribution";
+      return signalPassFailure();
+    }
+
+    auto maybeSubgroupSize = getSubgroupSize(func);
+    if (!maybeSubgroupSize) {
+      func.emitError() << "subgroup size required for vector distribution";
+      return signalPassFailure();
+    }
+
+    OpBuilder builder(func);
+    builder.setInsertionPointToStart(&func.getFunctionBody().front());
+    SmallVector<OpFoldResult> threadGrid = {
+        builder.createOrFold<gpu::ThreadIdOp>(func.getLoc(), gpu::Dimension::x),
+        builder.createOrFold<gpu::ThreadIdOp>(func.getLoc(), gpu::Dimension::y),
+        builder.createOrFold<gpu::ThreadIdOp>(func.getLoc(),
+                                              gpu::Dimension::z)};
+    auto workgroupSize = getWorkgroupSize(func);
+    AffineExpr x, y, z;
+    bindSymbols(func.getContext(), x, y, z);
+    // Construct the expression for linearizing the thread indices.
+    AffineExpr linearId =
+        x + workgroupSize[0] * y + workgroupSize[1] * workgroupSize[0] * z;
+    AffineExpr laneId = linearId % *maybeSubgroupSize;
+
+    // This all needs some kind of simplification; the arithmetic it produces
+    // doest not get folded away as nicely as it could.
+    AffineMap idMap = AffineMap::getMultiDimIdentityMap(2, func.getContext());
+
+    // Clamp the thread indices to the workgroup sizes.
+    OpFoldResult c0 =
+        builder.createOrFold<arith::ConstantIndexOp>(func.getLoc(), 0);
+    threadGrid[0] = affine::makeComposedFoldedAffineMax(
+        builder, func.getLoc(), idMap, {threadGrid[0], c0});
+    threadGrid[1] = affine::makeComposedFoldedAffineMax(
+        builder, func.getLoc(), idMap, {threadGrid[1], c0});
+    threadGrid[2] = affine::makeComposedFoldedAffineMax(
+        builder, func.getLoc(), idMap, {threadGrid[2], c0});
+
+    OpFoldResult dimX = builder.createOrFold<arith::ConstantIndexOp>(
```

Comment:
I dont think `createOrFold` does anything here. You dont need this. You just need `builder.getIndexAttr(...)` ?